minidump-processor/src/symbols.rs

```diff
@@ -1,3 +1,98 @@
+//! This module defines the interface between minidump-processor and its [Symbolizer][].
+//!
+//! There can only be one [Symbolizer][], and this is configured by minidump-processor's Cargo
+//! feature flags. The currently defined Symbolizers are:
+//!
+//! * breakpad_symbols -- feature: breakpad-syms (currently the default)
+//! * symbolic -- feature: symbolic-syms (not yet implemented, but compiles)
+//!
+//!
+//! minidump-processor and the [Symbolizer][] communicate using a series of traits. The symbolizer
+//! must provide implementations of these traits:
+//!
+//! * [SymbolProvider][] - provides symbolication, cfi evaluation, and debug statistics
+//!     * Implemented by [Symbolizer][]
+//!     * This actually doesn't need to be a trait in the current design, it exists to allow
+//!       multiple symbolicators to be used together, via [MultiSymbolProvider][]. The other
+//!       SymbolProviders have been removed, but I figured it would be a waste to throw out
+//!       this minimally intrusive machinery.
+//!
+//! * [SymbolSupplier][] - maps a [Module][] to a [SymbolFile][]
+//!     * minidump-processor does not directly use this, it's just there so the Symbolizer can
+//!       generically handle different symbol fetching strategies (which minidump-processor
+//!       selects and configures).
+//!
+//!
+//!
+//! While minidump-processor provides implementations of these traits:
+//!
+//! * [FrameSymbolizer][] - callbacks that symbolication uses to return its results.
```

Comment:
I implemented this this way with an eye towards interop, not forcing callers of the symbolication APIs to choose between exposing `StackFrame` in their public API or writing code to transform a `StackFrame` into their own data type. I guess this isn't a huge improvement over that, but decoupling things slightly felt sensible. IDK if any of the Sentry crates have a better approach, or a trait we could put somewhere that'd be amenable to sharing?

Reply:
To me the primary purpose of these traits is to avoid a circular dependency between breakpad-symbols and minidump-processor. breakpad-symbols defines the traits so that it doesn't need to know about the concrete types like StackFrame at all.

Although it also is just genuinely being used to allow me to have DummyFrame in the stack scanner (although arguably we should just have a has_function_name method on SymbolProvider or whatever, but then I'd need to write that out like 5 times blech).